src/main/java/org/mockbukkit/mockbukkit/entity/AbstractVillagerMock.java

```diff
@@ -0,0 +1,116 @@
+package org.mockbukkit.mockbukkit.entity;
+
+import org.bukkit.entity.AbstractVillager;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+import org.bukkit.inventory.Merchant;
+import org.bukkit.inventory.MerchantRecipe;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.mockbukkit.mockbukkit.ServerMock;
+import org.mockbukkit.mockbukkit.inventory.MerchantInventoryMock;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Mock implementation of an {@link AbstractVillager}.
+ *
+ * @see AgeableMock
+ */
+public abstract class AbstractVillagerMock extends AgeableMock implements Merchant, AbstractVillager, InventoryHolder
+{
+	private final MerchantInventoryMock inventory = new MerchantInventoryMock(this, this);
+	private final List<MerchantRecipe> recipes = new ArrayList<>();
+
+	private @Nullable HumanEntity trader;
+
+	/**
+	 * Constructs a new {@link AbstractVillagerMock} on the provided {@link ServerMock} with a specified {@link UUID}.
+	 *
+	 * @param server The server to create the entity on.
+	 * @param uuid   The UUID of the entity.
+	 */
+	protected AbstractVillagerMock(@NotNull ServerMock server, @NotNull UUID uuid)
+	{
+		super(server, uuid);
+
+		updateTrades();
+	}
+
+	@Override
+	public @NotNull MerchantInventoryMock getInventory()
+	{
+		return this.inventory;
+	}
+
+	@Override
+	public void resetOffers()
+	{
+		this.recipes.clear();
+	}
+
+	@Override
+	public @NotNull List<MerchantRecipe> getRecipes()
+	{
+		return Collections.unmodifiableList(this.recipes);
```

Comment:
Should probably deep clone the list here. I'm having difficulties interpreting the implementation though

Reply:
I had the same doubt, but it seems like that in paper we are only creating a Immutable list of trades.
I'll try to break it down:

1. In this function the NMS Merchant Offers is converted to CraftMerchant, but it is only created once;
```java
public CraftMerchantRecipe asBukkit() {
  return (this.bukkitHandle == null) ? this.bukkitHandle = new CraftMerchantRecipe(this) : this.bukkitHandle;
}
```
2. In this function we are transforming/mapping the merchant offer in Merchant recipe by just calling the function from step 1 and then creating an immutable list.
```java
@Override
default List<MerchantRecipe> getRecipes() {
    return List.copyOf(Lists.transform(this.getMerchant().getOffers(), new Function<>() {
        @Override
        public MerchantRecipe apply(net.minecraft.world.item.trading.MerchantOffer recipe) {
            return recipe.asBukkit();
        }
    }));
}
```

So in the end it seems like that we are simply returning the original reference and not a clone.

Reply:
Okay, then that explains that. Might be a bug on the paper end if this is the case. As generally all of these operations should effectively deep clone.